lib/Value_Type.thy

```diff
@@ -0,0 +1,120 @@
+  | typ_of_num (Const (\<^const_name>\<open>Num.Bit1\<close>, _) $ t) = Type (@{type_name "bit1"}, [typ_of_num t])
+  | typ_of_num (Const (\<^const_name>\<open>Num.One\<close>, _))      = Type (@{type_name "num1"}, [])
+  | typ_of_num (Const (\<^const_name>\<open>Groups.zero\<close>, _))  = Type (@{type_name "num0"}, [])
+  | typ_of_num t = raise TERM ("type_of_num: number expected", [t]);
+
+fun force_nat_numeral (Const (@{const_name numeral}, Type ("fun", [num, _])) $ n) =
+      Const (@{const_name numeral}, Type ("fun", [num, @{typ nat}])) $ n
+  | force_nat_numeral (Const (@{const_name "Groups.one"}, _))  = @{term "1::nat"}
+  | force_nat_numeral (Const (@{const_name "Groups.zero"}, _)) = @{term "0::nat"}
+  | force_nat_numeral t = raise TERM ("force_nat_numeral: number expected", [t])
+
+fun make_type binding v lthy =
+  let
+    val n = case get_term_numeral v of
+              SOME n => n
+            | NONE => raise TERM ("make_type: numeral expected", [v])
+    val typ = typ_of_num n
+  in
+    lthy |> Typedecl.abbrev (binding, [], Mixfix.NoSyn) typ |> #2
+  end
+
+fun make_def binding v lthy =
+  let
+    val mk_eq = HOLogic.mk_Trueprop o HOLogic.mk_eq
+    val def_t = mk_eq (Free (Binding.name_of binding, @{typ nat}), force_nat_numeral v)
+  in
+    lthy |> Specification.definition NONE [] [] (Binding.empty_atts, def_t) |> #2
+  end
+
+in
+
+fun value_type_cmd no_def binding t lthy =
+  let
+    val t' = Syntax.read_term lthy t
+    val v = case get_term_numeral t' of SOME _ => t' | NONE => Value_Command.value lthy t'
+  in
+    lthy
+    |> make_type binding v
+    |> (if no_def then I else make_def binding v)
+  end
+
+val no_def_option =
+  Scan.optional
+    (Parse.$$$ "(" |-- Parse.!!! (Parse.reserved "no_def" >> K true) --| Parse.$$$ ")")
+    false;
+
+val _ =
+  Outer_Syntax.local_theory \<^command_keyword>\<open>value_type\<close>
+    "Declare a type synonym from a term that evaluates to a numeral"
+    (no_def_option -- Parse.binding -- (\<^keyword>\<open>=\<close> |-- Parse.term)
+      >> (fn ((no_def, n), t) => value_type_cmd no_def n t))
+
+end
+\<close>
+
+(* These need to be outside "experiment" to work *)
+definition num_domains :: int where
+  "num_domains = 16"
+
+definition num_prio :: int where
+  "num_prio = 256"
+
+(* The rest works in local contexts: *)
+experiment
+begin
+
+text "The RHS does not have to be of type nat, it just has to evaluate to any numeral:"
+value_type num_queues = "num_prio * num_domains"
+
+text "This produces a type of the specified size and a constant of type nat:"
+typ num_queues
+term num_queues
+thm num_queues_def
+
+text "You can leave out the constant definition, and just define the type:"
+value_type (no_def) num_something = "10 * num_domains"
+
+typ num_something
+
+end
+
+
+(* Remove the demo definitions: *)
+hide_const num_domains
+hide_fact num_domains_def
+
+hide_const num_prio
+hide_fact num_prio_def
```

Comment:
I had suspected something like this. Thoughts, @lsf37 ?

Reply:
Yeah, @corlewis is right, we should put this in a separate theory. I had hoped to be able to pull the definitions into the `experiment` block, but while the definitions do work there, the code generator seems to refuse to generate anything for them inside the local context, which means that `eval` won't work.

In more detail: `eval` complains that there are not code equations for `num_domains` in the example if you move the definition into the local context, but if you try and explicit `export_code` it will already complain that there is no constant `num_domains`. This does make some sense, because if there are assumptions or fixed variables in the context the generated code will not have the right semantics.

All this means this is a limitation of the `value_type` command in general: it can't use terms on the rhs that are only defined in a local context. If we ever need that, we'll need to use something else to compute the numeral.

I'll add a comment to that effect and move the example into a separate theory.